prov/util/src/util_cmap.c

```diff
@@ -135,16 +135,52 @@ void ofi_cmap_del_handle(struct util_cmap_handle *handle)
 	cmap->release(&cmap->lock);
 }
 
+static inline int
+util_cmap_check_and_realloc_handles_table(struct util_cmap *cmap, fi_addr_t fi_addr)
+{
+	int ret = FI_SUCCESS;
+
+	if (OFI_LIKELY(fi_addr < cmap->num_allocated))
+		return ret;
+
+	while (fi_addr >= cmap->num_allocated) {
+		struct util_cmap_handle **handles_av_new =
+				realloc(cmap->handles_av,
+					(cmap->av->count +
+					 cmap->num_allocated) *
+					sizeof(*cmap->handles_av));
+		if (OFI_LIKELY(*handles_av_new != NULL)) {
+			cmap->handles_av = cmap->handles_av;
+			memset(&cmap->handles_av[cmap->num_allocated],
+			       0, sizeof(*cmap->handles_av) *
+				  cmap->av->count);
+			cmap->num_allocated += cmap->av->count;
+		} else {
+			ret = -FI_ENOMEM;
+			break;
+		}
+	}
+
+	return ret;
+}
+
 /* Caller must hold cmap->lock */
 int util_cmap_alloc_handle(struct util_cmap *cmap, fi_addr_t fi_addr,
 			   enum util_cmap_state state,
 			   struct util_cmap_handle **handle)
 {
+	int ret;
+
 	*handle = cmap->attr.alloc(cmap);
 	if (OFI_UNLIKELY(!*handle))
 		return -FI_ENOMEM;
 	FI_DBG(cmap->av->prov, FI_LOG_EP_CTRL, "Allocated handle: %p for "
 	       "fi_addr: %" PRIu64 "\n", *handle, fi_addr);
+	ret = util_cmap_check_and_realloc_handles_table(cmap, fi_addr);
+	if (OFI_UNLIKELY(ret)) {
+		cmap->attr.free(*handle);
+		return ret;
+	}
```

Comment:
How about checking for realloc on AV insert instead of here? I think that can avoid checking for realloc in `util_cmap_move_handle` as well. I think it also ensures that user is not realloc'ing cmap without increasing the AV size first.